### ISSUE TITLE
Indefinite Hang on `nrt_load_collectives`

### ISSUE THREAD

**Issue by xanderdunn** [aws-neuron/aws-neuron-sdk#699]
We're running a transformer with pipeline parallelsim across 4 trn1n.32xlarge instances, each with only 2 ranks for the sake of testing.

At some point early in our training we reach a situation where both of the ranks on the third instance hang indefinitely on the call to `nrt_load_collectives` for one of our models. 

We're familiar with cases where `nrt_execute` might hang indefinitely waiting on comms to happen, as described [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-runtime/nrt-troubleshoot.html?highlight=hang#communication-timeout), but that's not the scenario we're in here. We're finding that the hang is on `nrt_load_collectives`, and we would expect at least some kind of error from this call.

I captured the NRT DEBUG logs [here](https://gist.github.com/xanderdunn/a52be3c14e42d6340dacd0002c3c4ad9). 
Near the end of the logs you'll see the two models that are being loaded:
```
NRT:nrt_load_collectives                    API:IN: (start_nc=0, nc_count=1, global_device_id=4, global_device_count=8)
NRT:nrt_load                                API:IN: (start_nc=0, nc_count=1, name=/tmp/run_forward_rank_4_14004085392876107508.neff)
```
[here](https://gist.github.com/xanderdunn/a52be3c14e42d6340dacd0002c3c4ad9#file-nrt_debug_logs-txt-L5209).

And:
```
NRT:nrt_load_collectives                    API:IN: (start_nc=0, nc_count=1, global_device_id=5, global_device_count=8)
NRT:nrt_load                                API:IN: (start_nc=0, nc_count=1, name=/tmp/run_forward_rank_5_14697962644056566318.neff)
```
[here](https://gist.github.com/xanderdunn/a52be3c14e42d6340dacd0002c3c4ad9#file-nrt_debug_logs-txt-L5488).

`start_nc=0, nc_count=1` is correct for all ranks because we are setting `NEURON_RT_VISIBLE_CORES`. On the first one we have `NEURON_RT_VISIBLE_CORES=0` and on the second rank we have `NEURON_RT_VISIBLE_CORES=1`

In our code we have:
```rust
                info!("Have distributed_run_info {:?}", distributed_run_info);
                let result = nrt::nrt_load_collectives(
                    neff_data.as_ptr() as *const _,
                    neff_data.len(),
                    // Neuron core index to start from; since we set visible cores in NEURON_RT_VISIBLE_CORES
                    // it should be 0 if we run one rank per process.
                    (distributed_run_info.world_rank_id % num_ranks_per_process) as i32,
                    // Number of neuron cores to allocate the model rank to. We always use 1 core per rank.
                    1,
                    distributed_run_info.world_rank_id,
                    distributed_run_info.world_num_ranks,
                    model,
                );
                info!("Loaded model");
```
We find that the first `info!()` is printed, but the second one is never reached.

Are there known scenarios where `nrt_load_collectives` might hang? Is there any information in the DEBUG logs that might explain it? If necessary we can come up with a minimal reproducing sample here, but it will take considerable effort.

This is not specific to any particular instance, this has 100% reproducibility on any set of instances. 

```
$ neuronx-cc --version
NeuronX Compiler version 2.7.0.40+f7c6cf2a3

Running on AMI ami-01257e71ecb2f431c
Running in region usw2-az4
Running on python 3.8.10

HWM version 2.7.0.3-0092b9d34
NumPy version 1.21.6

$ dpkg -l | grep -i neuron
ii  aws-neuronx-collectives                2.14.9.0-f3942f597                amd64        neuron_ccom built using CMake
ii  aws-neuronx-dkms                       2.10.11.0                         amd64        aws-neuronx driver in DKMS format.
ii  aws-neuronx-oci-hook                   2.2.9.0                           amd64        neuron_oci_hook built using CMake
ii  aws-neuronx-runtime-lib                2.14.8.0-edaf1b171                amd64        neuron_runtime built using CMake
ii  aws-neuronx-tools                      2.11.10.0                         amd64        Neuron profile and debug tools
```

**Comment by xanderdunn** [comment]
@arktoswb pointed out that it is expected behavior that `nrt_load_collectives` will hang waiting for other ranks to make their call to `nrt_load_collectives`. We're trying to load once and then `nrt_execute` multiple times on multiple models, each with comms. 

It turns out, we were trying to have one instance of a model on rank n communicate with two different models on rank m. We consolidated so that we have a 1:1 model:model comms relationship, and that gets us past our hang. Still some issues to work through, will open a new issue if we have something more concrete.
